### ISSUE TITLE
Ability to mutate Request's CacheHeaders with data contained in ApolloResponse

### ISSUE THREAD

**Issue by ArjanSM** [apollographql/apollo-kotlin#3909]
**Use case**
As a part of our project to capture Client Side Cache Usage data, I want to log data associated with Cache Write events on Android Clients.
The expectation is to log cache_size along with some other metadata (`request_id` and `document_id`) whenever a Cache Write event occurs.
I've created a wrapper for the `NormalizedCache` which could help with the logging like mentioned below:-
```
class LoggingCache(val delegate: NormalizedCache): NormalizedCache() {
  override fun merge(record: Record, cacheHeaders: CacheHeaders): Set<String> {
    // Do your logging here
    return delegate.merge(record, cacheHeaders)
  }

  override fun merge(records: Collection<Record>, cacheHeaders: CacheHeaders): Set<String> {
    // Do your logging here
    return delegate.merge(records, cacheHeaders)
  }

  // more methods...
}
```
The `request_id` is fetched from the Server and while it can be intercepted via an `ApolloInterceptor` between the `ApolloCacheInterceptor` & `NetworkInterceptor` there isn't a convenient way to pass the `request_id` for the logging event.
The ability to mutate the `CacheHeaders` associated with the `ApolloRequest` would make life a lot easier.
**Describe the solution you'd like**
It would be awesome if there's an API that allows mutating the `CacheHeader` values after recieving the response from the Server. For e.g. `response.cacheHeaders`

1.  The easiest, definitely not the best :-), option would be to expose this function in the `ClientCacheExtension.kt`
```
internal val <D : Operation.Data> ApolloRequest<D>.cacheHeaders
  get() = executionContext[CacheHeadersContext]?.value ?: CacheHeaders.NONE
``` 
2. [Copy the `request.executionContext` to the `response.executionContext`](https://community.apollographql.com/t/cache-write-events-on-android/2912)

**Comment by BoD** [comment]
To avoid copying the request.executionContext to the response.executionContext, a lower impact change could be:

1. add `ApolloResponse.Builder<D>.cacheHeaders(cacheHeaders: CacheHeaders)` so interceptors can add their custom headers to a response
2. in `ApolloCacheInterceptor`, when calling `store.writeOperation`, pass a merging of the request's CacheHeaders **+** the response's one

WDYT?

**Comment by ArjanSM** [comment]
Could I take this?

**Comment by BoD** [comment]
Ooh yes @ArjanSM if you want to have a go at a PR this is much welcome!

**Comment by ArjanSM** [comment]
@BoD quick question. While merging the `ApolloRequest`'s and `ApolloResponse`'s `CacheHeaders`, don't we have the risk of overwritting the values of certain headers?

**Comment by BoD** [comment]
Sorry for the late reply - I [replied](https://github.com/apollographql/apollo-kotlin/pull/4102#issuecomment-1125746047) in your PR.

**Comment by ArjanSM** [comment]
Do I close this issue then? or wait for the release?

**Comment by martinbonnin** [comment]
Usually we set a "Fixed in SNAPSHOTs" label, which I just did! Thanks for the reminder 🙏

**Comment by BoD** [comment]
Change available in v3.3.1.
